### ISSUE TITLE
Collocation exact Hessian is CPU-dispatch-bound on GPU: capture it in a CUDA graph

### ISSUE THREAD

**Issue by JBjoernskov** [JBjoernskov/Twin4Build#122]
## Summary

The exact-Hessian collocation callback is **CPU-dispatch-bound, not GPU-bound**. Capturing it in a CUDA graph and replaying it should remove nearly all of that dispatch cost, with no change to the numerics.

## Evidence

Measured on an A100-SXM4-40GB (Xeon @ 2.20 GHz, torch 2.11.0+cu128) with `gpu_profile_hessian_kernels.ipynb`, added in #121.

**Exact is Gauss-Newton plus one call**, and that call dominates:

| cuda, maxiter=100 | gauss-newton | exact |
|---|---|---|
| obj + grad + constraints | 27.8 s | 27.7 s |
| constraint Jacobian | 9.5 s | 9.6 s |
| **Hessian** | 0.5 s | **177.1 s** |
| IPOPT KKT (CPU) | 14.5 s | 19.4 s |
| **total** | **52.3 s** | **233.9 s** |

Everything except the Hessian matches to within noise. The Hessian is **76% of wall-clock**.

**Per Hessian call:**

| | |
|---|---|
| CPU dispatch | **1259 ms** |
| GPU compute | **156 ms** |
| GPU busy fraction | **12.4%** |

The exact Hessian achieves **12.6%** of its Amdahl ceiling, and the GPU is busy **12.4%** of the call. Those being the same number is the point: the shortfall *is* the idle fraction.

**The GPU work itself is healthy** — 58.5% of device time is fp64 GEMM on tensor cores (`cutlass_80_tensorop_d884gemm`), in 88–140 µs kernels. There is simply too little of it relative to ~18,000 ATen dispatches per call, which come from `_expm_ss` unrolling ~26 matmuls (used under functorch because `matrix_exp` has no vmap batching rule) that `jacfwd(jacrev(·))` then differentiates twice.

## Ruled out

- **fp64 throughput** — measured on an A100 (fp64 at 1/2 of fp32, not a consumer card's 1/32).
- **Unsynchronised CUDA timing** — `torch_frac` moves by 0.001 with explicit `synchronize()`; the device-to-host copy in each callback was already forcing a sync.
- **`torch.compile`** — Dynamo cannot trace `vmap(jacfwd(jacrev(·)))`. It fails on doubly-wrapped `GradTrackingTensor(BatchedTensor(...))` even with `backend="eager"`, so this is a tracing limitation, not a missing compiler.
- **vmap chunking** — real but bounded, and already addressed in #121 via `TWIN4BUILD_HESS_CHUNK_DIV` (1.95× on the Hessian call). Not enough on its own.

## Proposed change

Capture the Hessian's **tensor region** in a CUDA graph and replay it per call.

Scope deliberately excludes the numpy boundary and the derivative cache:

| stays on the host | inside the graph |
|---|---|
| numpy → torch conversion | `_curvature` / `_obj_curvature` vmaps |
| `_derivs(z)` cache lookup | einsum accumulation into `Btt`/`Bty`/`Byy` |
| packing → numpy for CasADi | packing into the `vals` tensor |

The cached `d["Jt"]`, `d["Jx"]` are already computed by the Jacobian callback at the same `z`, so they enter as graph *inputs* rather than being recomputed.

### Required code change

`if s_gn != 0.0:` in `hess_vals_fn` is a **host branch on a value that changes** (`sigma` is 0 during IPOPT restoration), and a graph records one path. Make `s_gn` a 0-dim device tensor and multiply unconditionally. Costs the Gauss-Newton term when `sigma == 0`; buys a single graph and no host branch.

### Mechanism

1. Static input/output buffers allocated once.
2. Warm up ~3 calls on a side stream.
3. Capture with `torch.cuda.graph`.
4. Per call: `copy_` inputs into the static buffers, `replay()`, read the static output.

### Safety

- First replay asserted **bit-identical** to an eager evaluation at the same inputs; capture is disabled if it is not. Replay executes the identical recorded kernels, so this is an invariant rather than a tolerance.
- Any capture failure falls back to eager permanently, with one warning.
- `TWIN4BUILD_CUDA_GRAPH=0` to disable.
- The graph lives in the `_solve_sparse_collocation` closure, so it is naturally per-solve and cannot outlive a shape change.

## Expected payoff

Removing ~1.1 s of the 1.26 s dispatch, on a call that is 76% of wall-clock, should take the exact arm at `maxiter=100` from ~234 s toward **~80 s** (roughly 3×), stacking with the 1.95× already available from `HESS_CHUNK_DIV=1`.

**GPU only** — no CPU benefit, and CPU single-shooting remains the recommended path for calibration work (711 s, pooled 30.58 in #121's benchmark).

## Risks

- Any host sync inside the captured region (`.item()`, `.cpu()`, a Python branch on tensor values) breaks capture. The `sigma` branch above is the known one; others would surface as capture failures, which fall back safely.
- The graph holds a private memory pool (~0.3 GB at this problem size) for its lifetime.
- The payoff estimate assumes replay drives dispatch to near zero. That is the normal outcome for this signature but is not yet measured here.

## Follow-up, not in scope

The constraint Jacobian very likely has the same 8:1 dispatch:compute ratio and could reuse the wrapper. The profile that would have confirmed it was void — the replay hit the `_derivs` cache (keyed on `z.tobytes()`) and measured nothing. Worth a separate measurement with a cache-busting input.

**Comment by JBjoernskov** [comment]
## Result: CUDA graphs are not viable for this region

Measured on an A100-SXM4-40GB with `gpu_graph_capture_diagnostic.ipynb`. **Capture fails, so the proposal in this issue does not work as written.**

```
canary: OK -- the context can capture, so failures below are real

  curvature (constraints only)       global   FAIL  cudaErrorStreamCaptureInvalidated
  curvature (constraints only)       relaxed  FAIL  cudaErrorStreamCaptureInvalidated
  obj_curvature (+ composer.F_aug)   global   FAIL  cudaErrorStreamCaptureInvalidated
  obj_curvature (+ composer.F_aug)   relaxed  FAIL  cudaErrorStreamCaptureInvalidated
  hess_core (everything)             global   FAIL  cudaErrorStreamCaptureInvalidated
  hess_core (everything)             relaxed  FAIL  cudaErrorStreamCaptureInvalidated
```

A trivial `a @ a` capture was checked before the first attempt and after every one, and passed throughout — so these are six real, independent failures rather than one poisoned context echoing. (An earlier version of the diagnostic *did* have that flaw: it ran the solve with the graph enabled, which failed first and invalidated everything after it. Its confident "neither half captures" verdict was an artefact, and the run above is the corrected one.)

### Why

The simplest piece — `_curvature` — is `vmap(torch.func.hessian(...))` with no composer, no einsums and no packing. Its only extra machinery is the transform stack itself, and **`jacrev` runs a backward pass: `torch.autograd` executes inside the captured region.** The autograd engine runs on its own threads and synchronises, which invalidates stream capture.

That is precisely why PyTorch ships `torch.cuda.make_graphed_callables` for graphing backward — and that helper operates on `nn.Module`s with a conventional forward/backward, not on nested `torch.func` transforms.

`capture_error_mode="relaxed"` does not help, so the cause is not a tolerable synchronisation.

### What would be required

Removing autograd from inside the captured region — i.e. a hand-written Hessian instead of `jacfwd(jacrev(·))`. That is a far larger change than this issue proposed, and the payoff is capped near **3x** anyway: the Hessian is 76% of wall-clock and its dispatch ~89% of its own cost. Not obviously worth it.

### What was salvaged

The chunking fix, which is independent of the graph and is the real win:

`_hchunk = _deriv_chunk // 2` halved the Hessian's vmap **unconditionally**. On this problem the whole Hessian needs 0.30 GB against a 2 GB budget, so the Jacobian ran as one vmap while the Hessian ran as two — and `TWIN4BUILD_DERIV_BYTES` could not lift it, because `_deriv_chunk` is capped by segment count rather than budget. It is now sized from the Hessian's own budget.

Measured on the A100 (exact Hessian, 300 iterations):

| | seconds | pooled |
|---|---|---|
| cpu, old chunking | 1300.7 | 30.78 |
| cpu, new | 1081.3 | 30.22 |
| cuda, old chunking | 774.2 | 32.91 |
| **cuda, new** | **610.4** | **29.53** |

**`cuda` exact is now the first collocation arm to beat CPU single-shooting on both axes** — 610 s against 711 s, pooled 29.53 against 30.58. That is the containment argument finally delivering: a converged collocation must reach an equal-or-better objective, and every earlier arm was failing the "converged" half.

Verified bit-identical to `dev` across all 35,658 Hessian entries for the refactor. Note the chunking itself is *mathematically* exact but not *bitwise* reproducible across chunk sizes — BLAS picks different kernel tilings for batch 180 vs 360 — so the pooled differences above are real trajectory divergence over 300 interior-point iterations, not noise.

### Recommendation

Keep the chunk sizing. Drop the CUDA-graph machinery, or leave it default-off with this finding recorded, in case a future torch makes `torch.func` capturable.

**Comment by JBjoernskov** [comment]
The original capture attempt documented here was blocked by tensor-to-Python structural branches, but PR #127 removed those branches through static support contracts and successfully shipped direct CUDA Graph replay for exact Hessians. The PR measured a 3.39x Hessian-path speedup with identical converged results. Closing as completed by the later implementation.
